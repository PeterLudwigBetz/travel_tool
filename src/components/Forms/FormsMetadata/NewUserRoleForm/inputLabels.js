const inputLabels = {
  email: {
    label: 'Andela Email Address',
  },
  roleName: {
    label: 'Role'
  },
};

export default inputLabels;
