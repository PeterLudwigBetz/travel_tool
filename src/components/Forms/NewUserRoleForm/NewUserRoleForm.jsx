import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
import { FormContext } from '../FormsAPI';
import PersonalDetailsFieldset from './FormFieldsets/PersonalDetails';
import SubmitArea from '../NewRequestForm/FormFieldsets/SubmitArea';

class NewUserRoleForm extends PureComponent {
  defaultState = {
    values: {
      email: '',
      roleName: ''
    },
    errors: {},
    hasBlankFields: true
  };

  state = { ...this.defaultState };

  componentWillUnmount() {
    const { getRoleData } = this.props;
    this.handleCancel();
    getRoleData();
  }

  handleSubmit = event => {
    event.preventDefault();
    const { handleUpdateRole } = this.props;
    const { values } = this.state;
    if (this.validate()) {
      let data = values;
      handleUpdateRole(data);
    }
  };

  handleCancel = () => {
    this.setState({ ...this.defaultState });
  };

  validate = field => {
    let { values, errors } = this.state;
    [errors, values] = [{ ...errors }, { ...values }];
    let hasBlankFields = false;
    !values[field]
      ? (errors[field] = 'This field is required')
      : (errors[field] = '');

    hasBlankFields = Object.keys(values).some(key => !values[key]);
    this.setState(prevState => {
      return { ...prevState, errors, hasBlankFields };
    });
    return !hasBlankFields;
  };

  render() {
    const { values, errors, hasBlankFields } = this.state;
    const { updatingRole } = this.props;
    return (
      <FormContext targetForm={this} errors={errors} validatorName="validate">
        {updatingRole && (
          <h5 style={{ display: 'flex', justifyContent: 'center', fontFamily: 'DIN Pro' }}
          >
          Updating role...
          </h5>
        )}
        <form onSubmit={this.handleSubmit} className="new-request">
          <PersonalDetailsFieldset values={values} />
          <hr />
          <SubmitArea
            onCancel={this.handleCancel}
            hasBlankFields={hasBlankFields}
            send="Submit"
          />
        </form>
      </FormContext>
    );
  }
}

NewUserRoleForm.propTypes = {
  handleUpdateRole: PropTypes.func.isRequired,
  updatingRole: PropTypes.bool,
  getRoleData: PropTypes.func.isRequired,
};

NewUserRoleForm.defaultProps = {
  updatingRole: false,
};

export default NewUserRoleForm;
