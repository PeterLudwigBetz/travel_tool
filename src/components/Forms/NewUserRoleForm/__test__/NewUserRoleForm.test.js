import React from 'react';
import moment from 'moment';
import NewUserRoleForm from '../NewUserRoleForm';

describe('<NewUserRoleForm />', () => {
  let wrapper, onSubmit;
  onSubmit = jest.fn();
  
  const props = {
    user: {
      UserInfo: {
        name: 'John Doe'
      },
    },
    loading: false,
    errors: [],
    getRoleData: jest.fn(() => {}),
    handleUpdateRole:  jest.fn(() => {}),
    onChange: jest.fn()
  };

  const defaultState = {
    values: {
      email: '',
      roleName: '',
    },
    errors: {},
    hasBlankFields: true
  };


  beforeEach(() => {
    wrapper = mount(<NewUserRoleForm {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders two fieldsets', () => {
    const fieldsets = wrapper.find('fieldset');
    expect(fieldsets).toHaveLength(2);
  });

  it('renders the personal details fieldset', () => {
    const personalDetails = wrapper.find('fieldset.personal-details');
    expect(personalDetails).toHaveLength(1);
  });

  it('picks input values', () => {
    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        email: 'email',
        value: 'Tomato Guy'
      }
    });
    expect(wrapper.state().values.email).toBe('Tomato Guy');
  });


  it('validates input on blur', () => {
    wrapper.find('input[name="email"]').simulate('blur');
    wrapper.update();
    expect(wrapper.state().errors.email).toBe('This field is required');
  });


  it('validates form before sending data', () => {
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  it('should render a loading indicator while updating a new role', () => {
    wrapper.setProps({ updatingRole: true });
    expect(wrapper.find('h5').text()).toEqual('Updating role...');
  });


  it('calls on submit when all details are correct', () => {
    wrapper.setState({
      values: {
        email: 'test',
        roleName: 'Requester',
      }
    });

    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit');
    expect(spy).toHaveBeenCalledTimes(1);
    wrapper.state().values.email = 'test';

    expect(props.handleUpdateRole).toHaveBeenCalledWith(wrapper.state().values);//eslint-disable-line
    expect(props.handleUpdateRole).toHaveBeenCalledTimes(1);//eslint-disable-line
  });


  it('should clear the form when the component unmounts', () => {
    const componentWillUnmount = jest.spyOn(wrapper.instance(), 'componentWillUnmount');
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalledTimes(1);
  });

});
