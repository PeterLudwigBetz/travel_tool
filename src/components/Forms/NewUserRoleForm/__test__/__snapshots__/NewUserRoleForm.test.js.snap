// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewUserRoleForm /> renders correctly 1`] = `
<form
  className="new-request"
  onSubmit={[Function]}
>
  <fieldset
    className="personal-details"
  >
    <div
      className="input-group"
    >
      <div
        style={
          Object {
            "marginRight": "30px",
            "width": " 232px",
          }
        }
      >
        <div
          className="form-input  "
        >
          <label
            htmlFor="email"
          >
            Andela Email Address
            <span
              style={
                Object {
                  "color": "red",
                }
              }
            >
              *
            </span>
          </label>
          <input
            className=" "
            label="Andela Email Address"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            size="232px"
            type="text"
            value=""
          />
          <span
            className="error"
          />
        </div>
      </div>
      <div
        className="form-input  "
      >
        <label
          htmlFor="roleName"
        >
          Role
          <span
            style={
              Object {
                "color": "red",
              }
            }
          >
            *
          </span>
        </label>
        <div
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            choices={
              Array [
                "Super Administrator",
                "Travel Administrator",
                "Travel Team Member",
                "Requester",
                "Manager",
              ]
            }
            className="input select  "
            error=""
            label="Role"
            name="roleName"
            onChange={[Function]}
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            size="232px"
            tabIndex="0"
            type="dropdown-select"
            value=""
          >
            <div
              className="value"
              style={
                Object {
                  "width": "232px",
                }
              }
            >
              <div
                className="select-dropdown inactive"
              >
                <img
                  alt="icn"
                  src="form_select_dropdown.svg"
                />
              </div>
            </div>
            <ul
              className="select-menu select-menu--inactive"
            >
              <li
                id="choice"
                key="Super Administrator"
              >
                <div
                  id="choice"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Super Administrator
                </div>
              </li>
              <li
                id="choice"
                key="Travel Administrator"
              >
                <div
                  id="choice"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Travel Administrator
                </div>
              </li>
              <li
                id="choice"
                key="Travel Team Member"
              >
                <div
                  id="choice"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Travel Team Member
                </div>
              </li>
              <li
                id="choice"
                key="Requester"
              >
                <div
                  id="choice"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Requester
                </div>
              </li>
              <li
                id="choice"
                key="Manager"
              >
                <div
                  id="choice"
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Manager
                </div>
              </li>
            </ul>
          </div>
        </div>
        <span
          className="error"
        />
      </div>
    </div>
  </fieldset>
  <hr />
  <fieldset>
    <div
      className="submit-area"
    >
      <button
        className="bg-btn bg-btn--inactive"
        id="cancel"
        onClick={[Function]}
        type="button"
      >
        Clear
      </button>
      <button
        className="bg-btn bg-btn--active"
        disabled={true}
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </div>
  </fieldset>
</form>
`;
